### PR TITLE
support complex delete expression

### PR DESCRIPTION
**Pull request by aoiasd**
/related https://github.com/milvus-io/milvus/issues/25751

### CODE REVIEW

**Comment by aoiasd** [comment]
rerun ut

File: internal/querynodev2/segments/retrieve.go [hunk 1]
```diff
@@ -81,23 +84,104 @@ func retrieveOnSegments(ctx context.Context, segments []Segment, segType Segment
...
+			}
+
+			errs[i] = nil
+			metrics.QueryNodeSQSegmentLatency.WithLabelValues(fmt.Sprint(paramtable.GetNodeID()),
+				metrics.QueryLabel, label).Observe(float64(tr.ElapseSpan().Milliseconds()))
+		}(segment, i)
+	}
+	wg.Wait()
+
+	for _, err := range errs {
```
Comment (jiaoew1991) [inline internal/querynodev2/segments/retrieve.go:129]: should combine the errors into one by merr.Combine

File: internal/types/types.go [hunk 2]
```diff
@@ -1448,7 +1450,9 @@ type QueryNode interface {
 	Search(ctx context.Context, req *querypb.SearchRequest) (*internalpb.SearchResults, error)
 	SearchSegments(ctx context.Context, req *querypb.SearchRequest) (*internalpb.SearchResults, error)
 	Query(ctx context.Context, req *querypb.QueryRequest) (*internalpb.RetrieveResults, error)
+	QueryStream(ctx context.Context, req *querypb.QueryRequest, streamer streamrpc.QueryStreamer) error
 	QuerySegments(ctx context.Context, req *querypb.QueryRequest) (*internalpb.RetrieveResults, error)
+	QuerySegmentsStream(ctx context.Context, req *querypb.QueryRequest, streamer streamrpc.QueryStreamer) error
```
Comment (jiaoew1991) [inline internal/types/types.go:1455]: rename to QueryStreamSegment, Looks more unified.

File: internal/querynodev2/cluster/worker.go [hunk 3]
```diff
@@ -139,6 +142,36 @@ func (w *remoteWorker) QuerySegments(ctx context.Context, req *querypb.QueryRequ
...
+	client := streamer.AsClient()
+	for {
+		result, err := client.Recv()
+		if err != nil {
+			return err
+		}
+
+		err = merr.Error(result.GetStatus())
+		if errors.Is(err, merr.ErrPartitionResult) {
+			err = srv.Send(result)
```
[the remark this answers is not in the document]
Comment (aoiasd) [reply internal/querynodev2/cluster/worker.go:161]: Every stream rpc call use a stream server one's self, so data won't be mixed up.

**Comment by aoiasd** [comment]
rerun ut

File: internal/proxy/task_delete.go [hunk 4]
```diff
@@ -234,102 +223,226 @@ func (dt *deleteTask) Execute(ctx context.Context) (err error) {
 	defer sp.End()
 	log := log.Ctx(ctx)
 
-	tr := timerecord.NewTimeRecorder(fmt.Sprintf("proxy execute delete %d", dt.ID()))
+	if len(dt.req.GetExpr()) == 0 {
+		log.Warn("empty expr")
```
Comment (xiaofan-luan) [inline internal/proxy/task_delete.go:227]: throw err when empty empty expr.

File: internal/proxy/task_delete.go [hunk 5]
```diff
@@ -234,102 +223,226 @@ func (dt *deleteTask) Execute(ctx context.Context) (err error) {
...
+		err := dt.simpleDelete(ctx, termExp, stream)
+		if err != nil {
+			return err
+		}
+	} else {
+		// if get complex delete expr
+		// need query from querynode before delete
+		err = dt.complexDelete(ctx, plan, stream)
+		if err != nil {
+			log.Warn("complex delete failed,but delete some data", zap.Int("count", dt.count))
```
Comment (xiaofan-luan) [inline internal/proxy/task_delete.go:254]: log the expression

File: internal/proxy/task_delete.go [hunk 6]
```diff
@@ -234,102 +223,226 @@ func (dt *deleteTask) Execute(ctx context.Context) (err error) {
...
+
+		client := streamer.AsClient()
+		for {
+			result, err := client.Recv()
+			if err != nil {
+				return err
+			}
+
+			err = merr.Error(result.GetStatus())
+			if errors.Is(err, merr.ErrPartitionResult) {
```
Comment (xiaofan-luan) [inline internal/proxy/task_delete.go:327]: is there a way we don't use Err to indicate partition result?

Comment (xiaofan-luan) [reply internal/proxy/task_delete.go:327]: because this might affect retry and other handling

File: internal/proxy/task_delete.go [hunk 7]
```diff
@@ -234,102 +223,226 @@ func (dt *deleteTask) Execute(ctx context.Context) (err error) {
...
+	}
+
+	return nil
+}
+
+func (dt *deleteTask) PostExecute(ctx context.Context) error {
+	return nil
+}
+
+func (dt *deleteTask) getQueryExecuteWithStream(stream msgstream.MsgStream, plan *planpb.PlanNode) executeFunc {
```
Comment (xiaofan-luan) [inline internal/proxy/task_delete.go:266]: change to a better naming.

For instance,
StreamingQueryAndDelte

File: internal/proxy/task_delete.go [hunk 8]
```diff
@@ -98,27 +111,20 @@ func (dt *deleteTask) getChannels() []pChan {
...
-	}
-
-	plan, err := createExprPlan(schema, expr)
-	if err != nil {
-		return res, 0, fmt.Errorf("failed to create expr plan, expr = %s", expr)
-	}
-
-	// delete request only support expr "id in [a, b]"
-	termExpr, ok := plan.Node.(*planpb.PlanNode_Predicates).Predicates.Expr.(*planpb.Expr_TermExpr)
+func checkPlan(plan *planpb.PlanNode) (bool, *planpb.Expr_TermExpr) {
```
Comment (xiaofan-luan) [inline internal/proxy/task_delete.go:114]: Raname to getExpr

File: internal/util/streamrpc/streamer.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
+package streamrpc
+
```
Comment (xiaofan-luan) [inline internal/util/streamrpc/streamer.go:2]: did we test the behaviour of 
1) what will streamer happen on balance?
2) how will streamer handle timeout?

File: internal/querynodev2/delegator/delegator.go [hunk 10]
```diff
@@ -268,6 +270,68 @@ func (sd *shardDelegator) Search(ctx context.Context, req *querypb.SearchRequest
...
+		zap.Int("growingNum", len(growing)),
+	)
+	tasks, err := organizeSubTask(req, sealed, growing, sd.workerManager, sd.modifyQueryRequest)
+	if err != nil {
+		log.Warn("query organizeSubTask failed", zap.Error(err))
+		return err
+	}
+
+	_, err = executeSubTasks(ctx, tasks, func(ctx context.Context, req *querypb.QueryRequest, worker cluster.Worker) (*internalpb.RetrieveResults, error) {
+		return nil, worker.QueryStreamSegments(ctx, req, srv)
```
Comment (xiaofan-luan) [inline internal/querynodev2/delegator/delegator.go:323]: is there a ganruatee that growing sealed segment won't be changed during the query?

Comment (xiaofan-luan) [reply internal/querynodev2/delegator/delegator.go:323]: and do we need the ganruatee?

**Comment by xiaofan-luan** [comment]
overall it looks good to me.
But I still has some concern about the work manager implementation and GRPC streaming client.
1. there seems to be client leakage if old querynode down.
2. it didn't handle error like cross cluster error and serverID mismatch error(might not need to be handle)
3. how did streaming rpc do retry? 
4. what happened if we are streaming while channel/segment balanced?

**Comment by xiaofan-luan** [comment]
I'm not a big fan to introduce streaming rpc unless it shows better performance than staless rpc

**Comment by aoiasd** [comment]
rerun ut

**Comment by aoiasd** [comment]
rerun ut

**Comment by aoiasd** [comment]
rerun ut

**Pull request merged**